Throttle limit feature

etracer65 opened the pull request:
Fixes #5567 

Adds new `throttle_limit_type` and `throttle_limit_percent` parameters that allow the pilot to limit the maximum commanded throttle seen by the flight controller and determine how the limiting is applied.  `throttle_limit_type` can be `OFF | SCALE | CLIP`.  The `SCALE` value will scale the commanded throttle linearly to the limit specified by `throttle_limit_percent` based on the full stick range.  The `CLIP` type will leave stick scaling unchanged below the limit, and cap the throttle so that additional throttle input above the limit has no effect.

The settings are part of the Rate Profile.

These capabilities replace methods of limiting throttle in the radio which some pilots are using to manage throttle on tight courses or reduce overall battery consumption when the extra power isn't needed.

There is no effect on the maximum throttle seen by the motors so the mixer still has full authority.

Flight tested in normal mode and bench tested in various 3D modes (normal and on-a-switch).

etracer65 commented:
Note that this will conflict with #5528 so based on the merge order the other will need to rebase/fix conflicts.  Might be better to wait on this PR if #5528 will be ready in a reasonable amount of time as that one has already had to rebase at least once.

And yes, the conditional around the calculation is not technically needed as the 100% value will still be mathematically correct.  However the 100% condition will be what's used by nearly all pilots so I wanted to avoid the floating point multiply and divide when not needed as the mixer is speed critical code.

theSkonk commented:
Is this associated with rate profiles?

ctzsnooze commented:
@etracer65  

Thanks for this!  I was looking forward to it.

I was thinking it would be a true 'limit', not a 'slope reduction'.   A 'sensitivity' or 'slope reduction' might affect normal throttle responsiveness and cause racers to crash.  

This implementation changes the 'slope' of the throttle line, which will reduce throttle responsiveness throughout the range.  My feeling is that a true *limit* on throttle would be better, with normal throttle responsiveness until the limit was reached.  i.e. throttle = min (limit, throttle).  This way normal flight would be relatively unaffected, only sustained full throttle runs, or when using full throttle in certain manoeuvres.

Now the latter bit is tricky, because a lot of fast turns and recoveries *do* need full throttle or you crash.  So I wondered how would might play out with @joelucid's throttle boost feature?  And I was thinking that it would be great if the throttle boost code came *after* the limiting code.  

That's be cause the boost always causes a little overshoot.  

If the throttle limit was a hard limit, some boost overshoot could be configured get a transient for hard quick throttle inputs so maybe up to 100% throttle (i.e. when pulling up hard to avoid a ground impact). 

Unfortunately throttle boost would increase overall throttle sensitivity to quick inputs and change how it felt.  But if the pilot had throttle boost on all the time, and were accustomed to it, the effect of any limit on quick inputs would be minimal.

So basically my thoughts are:

1.  Limit throttle max (don't change sensitivity/slope)
2.  Action the throttle boost code after the limit, allowing the boost effect to transiently (very transiently) exceed the 'limit'
3.  Of course, PIDs should get 100% throttle authority if required.

I'm wondering if we could discuss these issues a bit more?

mikeller commented:
@theSkonk: No, it is in the PID profile at the moment. But I think it might be something to go into rate profiles, which would make it switchable in-flight. @etracer65, what do you think?

ctzsnooze commented:
Both ways can be done in the radio.  The change slope approach is obvious, but the limit can be done as a true limit using a curve in the radio (normal 45 degree line that flattens off at the top).  

A true limit does result in a 'dead stick' above the max value, but you get a dead stick when you hit the max travel of the stick anyway, don't you?  :-)  So a true 'limit' value is like a mechanical end-stop limiting stick travel to (say) 95%. 

A true limit has the advantage of absolutely normal responsiveness except for when seeking full throttle.  

I think the intent of this kind of thing, for racers, is not to wreck how the quad flies most of the time, but to preserve battery wastage during full throttle long straights.  These drain the battery for little time gain.  To preserve battery so that they can finish the race.  

But I think that if possible, they don't want the instant and normal throttle responses to change otherwise.

One way to achieve this might be some kind of iTerm like accumulator that was only active above the limit.  As it accumulated, the limiting became more pronounced, so that after whatever time it took to accumulate full attenuation (say 0.5s), you then became throttle limited.  But as soon as the pilot dropped below the limit, this accumulator would go away, and the attenuation would go to zero.  So for the slower parts of the track the pilot would have full authority, including for quick throttle blips.  But on the fast part, sustained throttle would be backed off, hopefully not even noticeably.

I'm envisaging a sort of 'virtual battery droop', I guess.

mikeller commented:
@ctzsnooze: A couple of points:

>  but the limit can be done as a true limit using a curve in the radio

One of the next features planned to be added to Betaflight are rate curves, in a similar way that they are available now on transmitters. This will supersede the different rate formulae, but this will also be useful to set up a throttle curve that can provide throttle limiting.
So the feature implemented in this request will probably be superseded / integrated by the rate curves, and we should just decide if it should be 'percentage', 'deadband', or switchable between these two.

> I think the intent of this kind of thing, for racers, is not to wreck how the quad flies most of the time, but to preserve battery wastage during full throttle long straights. These drain the battery for little time gain.

I think this is an interesting consideration. But I am not sure if it can be automated like you suggest, because once you're on the last straight before the finish, and you want that last little bit of speed to get ahead of your opponent, you're not going to be wanting your quad to limit your power. Maybe we need a 'kompressor' switch?

I can see both being supported in parallel, a hard limit to protect components / provide 'training wheels' for flyers after upgrading to higher voltage batteries, and a 'dynamic limiter' like you suggest, to give racers a competitive advantage without having to master the skill of self limitation.

ctzsnooze commented:
Ah indeed self-limitation is a something most true racers find very difficult when the race is closing and there's someone just in front!  :-)  I've seen both pilots literally fall from the sky with empty batteries after racing too hard, only to have a much more conservative pilot cruise home.

We have the Aussie Open coming up, some of the quads we fly can run flat out for about 1:30 - enough for three laps of our typical circuits.  But then the battery just dies.  Usually our circuits are short and technical enough that we are doing 22-25s laps and have room to spare.  But the Aussie Open is in a bigger location than usual, so the organisers may fly it longer.  Adding a heavier battery or more efficient props changes the whole handling of a quad... and the pilots just can't fly instinctively and reflexively any more.  They all say they will just go a bit lighter on the throttle, take it a little easy, and sustain the battery to the end of the race.  But most pilots just can't do that. 80% throttle down the straights seems sort of impossible for most of them.  :-)

You can appreciate the appeal of something that allows the pilots to 'floor it' when they need to jump out, or down the straights, have it fly like normal in the technical slow parts, and just get an extra few seconds battery life because on the long straights they're not just pushing too hard for little gain...  :-)  

I'm too old to be competitive so I don't mind  :-)

As to making curves etc in betaflight I'm happy with what I have.  I used to make a lot of complex curves, given up on that now.  Clever things like throttle feed forward work wonders.

One thing that would improve this PR perhaps would be the ability to do the reverse, i.e. not just cut throttle, but boost it.  I say this because a lot of us have our fastest best quad, we like how that flies.  But we have some older ones and they just don't quite have the motor authority.  Well, these days we can get feed-forward on roll and pitch with D weight, and feed-forward throttle so that it becomes more responsive.  But it is sort of tough hovering at a higher overall stick position, compared to the regular quads.  So that is the only place I still use a throttle curve, and its a simple linear boost.  Sure I get a deadband at the top, but I get the overall responsiveness closer to my faster ones.

So a simple change maybe to this PR is to allow a range of 50 to 150 say. So people can have a rates setting that truly spruces up the old bird.

borisbstyle commented:
@ctzsnooze How would you implement 150% of throttle? That doesn't really make sense to me. This will just bring the system into saturation earlier, without any benefit and still truncate the additional throttle input leading into deadband in upper throttle and decrease of resolution.

ctzsnooze commented:
I have some quads with really good new motors, and some with less strong motors, and some with sunny sky 2204's which are now very old.  

For the weaker ones I can run more D weight to feed-forward the sticks.  This compensates, to some extent, in roll and pitch, for the reduced torque.  And run a bit more P and D to since i have overall less authority overall in roll and pitch.  With these changes they actually roll and turn quite OK.

For throttle, with @joelucid's throttle feed-forward, I can get that extra bit of boost to overcome the lack of torque on quick throttle blips, so they accelerate almost the same as a faster one without so much (or no) throttle boost.

But for absolute thrust vs throttle position, what can I do?  Either I must teach my fingers a different hover point for each of them, and move the throttle over much bigger distances when on my other quads it only needs some bit less.  Or, I can make a steeper throttle line.

I personally find it a lot easier to fly the weaker ones by configuring a 'positive angle' on throttle.  I set it so that they hover with my thumb in approximately the same position.  This also means, with feed forward particularly, that they accelerate *sort of* the same with normal stick inputs.

The drawback of the steeper throttle line is that it hits max throttle before the stick hits the top of its travel.  I don't mind.  Full throttle is full throttle.  I just have to be mindful that recovering from a fast dive will be more problematic because I lack top-end thrust.  

But otherwise it's a lot easier to have fun with an older machine this way.

I guess that's the rationale.  Of course the PIDs are not changed, the rest of the 'system' works like normal.  And my thumb on throttle just does what it normally does.  

In a sense this is the same argument I'm making for the throttle attenuation process.  I would not want it to change my general throttle functionality, the throttle sensitivity to stick movement, in the lower ⅔ of the stick, if at all possible.  

Maybe this PR might work better by keeping normal throttle response (1:1) until say ⅔ throttle, then starting the attenuation above that point.  Sort of like TPA with a start point and then the attenuation above that point.  And even better with a smooth curve at the top.  I have done this with my radio, it works nicely and is hardly noticeable.  But of course it's not dynamic, like my earlier suggestion, which would be the best way, by far.

ctzsnooze commented:
![untitled](https://user-images.githubusercontent.com/11737748/38245571-e4def27e-3781-11e8-91df-5d33311ce08b.jpeg)
picture speaks better than words

etracer65 commented:
@mikeller Regarding having this in PID profile vs. rate profile, I'm open to either solution.  The differentiating fact is whether changing while in flight is really needed.  Also calling it a "rate" feature may be stretching it a little.

@ctzsnooze It seems like were going down the rabbit hole here vs. what is a simple request.  Perhaps we extend the limiting slightly to support both modes; scaling and clipping.  So add a `throttle_limit_type` with `OFF | SCALE | CLIP`.  The "scale" mode works as currently coded and the "clip" mode simply sets a hard limit with no scaling to the curve below (and having "dead stick" above the limit).

Other more complex "curves", etc. should probably be left for future implementations.  Also there is another PR that looks to address voltage/throttle compensation (#5528 ).

ctzsnooze commented:
@etracer65 - no problem at all with keeping this one simple.

theSkonk commented:
@etracer65 the reason for having it on the 'rate' profile is to associate it with ones race rates. It's not somthing I'd want for acro flying but for racing and I like to fly both high and low rates depending on what I'm doing. Having it on a Switch doesn't worry me by I see how this found be useful

mikeller commented:
@etracer65: Agreed that the current naming for (PID) profiles / rate profiles isn't ideal, and we'll have to revisit. My approach is to add things that are or can be made switchable in flight, and where there are use cases for switching in flight, to rate profiles, to make them switchable.

Also, agreed re `throttle_limit_type` with `OFF | SCALE | CLIP`.

etracer65 commented:
Ok, I'll work on the changes.

ctzsnooze commented:
Just a bit off topic since we are keeping this simple, but here is a prototype in excel of a 'smart' throttle limit.  Time scale in ms.  Red line = throttle input, blue line = throttle output after (a) a little derivative feed forward, and (b) smart throttle limit with time constant about 200ms, such that after about 0.5s sustained full throttle, output cuts to 80%, the rest of the time throttle behaviour is totally normal, even quick blips to full throttle are largely unaffected:
 
![smart throttle limit](https://user-images.githubusercontent.com/11737748/38305836-b7099d74-3851-11e8-8bf3-094d65f4c737.jpg)

This is what I think racers would really appreciate; normal flight dynamics, access to full throttle authority for transients, but while holding full throttle the battery gets preserved.

The coding is a bit too complex for my liking - fun challenge for some, difficult for me  :-)

ctzsnooze commented:
Same system with faster time constant just to illustrate behaviour:

![smart throttle limit faster](https://user-images.githubusercontent.com/11737748/38306054-9a4b43a8-3852-11e8-96b2-c6ae4392d901.jpg)

etracer65 commented:
Updated to add `throttle_limit_type` offering both scaled and clipping type behavior.  Also moved the settings from PID Profile to Rate Profiles so the pilot can select different settings while in flight by changing rate profiles.

SteveKfpv commented:
Just a note @ctzsnooze I think the original option is still really helpful particularly for those wanting to try out 5-6S on their usual mid-high hv motors without pulling too many amps. I'm interested to try 5S on my 2550kv with the throttle scaled to about 85% and see how it goes

ctzsnooze commented:
@SteveKfpv - Any of the above options will have their uses.  

Someone going to higher cell count for freestyle, for example, but wanting similar throttle response to what they had with fewer cells, might choose the linear reduction, so that it accelerates sort of the same for a given throttle input.  Even if they do this, they will still need to re-tune their PIDs, because the PIDs will have much more power behind them with a higher voltage pack.  The alternative is to restrict stick output, not just throttle but all sticks, which can be done in the Tx.  That way you can run a higher cell count battery, need less current, but have exactly the same handling.

The other options are more for people who want to get the full benefit of more aggressive throttle and PID responses in normal flight, but don't want drain or overheat their batteries or burn out their ESCs during sustained full throttle runs.  i.e. for racers who turn up and find that they just can't quite complete the track on their normal setup because their batteries aren't quite making it.  That's what the 'clip' option is for, but the dynamic approach I've suggested will probably be even more transparent to the pilot for this application.  

It's exciting to have a range of throttle management options coming up.

brucesdad13 commented:
This is INCREDIBLE 😁
